File: 2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        long long int res = 0, currSum=0;
        unordered_set<int> s;
        for(int lt=0, rt=0; lt<=nums.size()-k; rt++){
            if (s.find(nums[rt])!=s.end())  // found
                lt=rt, rt--, s.clear(), currSum=0;
            else s.insert(nums[rt]), currSum+=nums[rt];
            if(s.size()==k)
                res=max(res, currSum), currSum-=nums[lt], s.erase(nums[lt]), lt++;
        }
        return res;
    }
};

File: 2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        Set<Integer> window = new HashSet<>();
        int n = nums.length;
        long sum = 0, ans = 0;
        for (int i = 0; i < n; i++) {
            System.out.println(i + "\t" + sum + "\t" + window);
            if (!window.contains(nums[i])) {
                if (window.size() < k) {
                    window.add(nums[i]);
                    sum += nums[i];
                } else {
                    window.add(nums[i]);
                    window.remove(nums[i - k]);
                    sum += nums[i] - nums[i - k];
                    ans = Math.max(sum, ans);
                }
            } else {
                while (!window.isEmpty() && window.contains(nums[i])) {
                    sum -= nums[i - window.size()];
                    window.remove(nums[i - window.size()]);
                }
            }
        }
        return ans;
    }
}